cmd/kubeadm/app/cmd/phases/upgrade/node/preflight.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/options"
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow"
+	"k8s.io/kubernetes/cmd/kubeadm/app/preflight"
+	utilsexec "k8s.io/utils/exec"
+)
+
+// NewPreflightPhase creates a kubeadm workflow phase that implements preflight checks for a new node join
+func NewPreflightPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "preflight",
+		Short: "Run upgrade node pre-flight checks",
+		Long:  "Run pre-flight checks for kubeadm upgrade node.",
+		Run:   runPreflight,
+		InheritFlags: []string{
+			options.IgnorePreflightErrors,
+		},
+	}
+}
+
+// runPreflight executes preflight checks logic.
+func runPreflight(c workflow.RunData) error {
+	data, ok := c.(Data)
+	if !ok {
+		return errors.New("preflight phase invoked with an invalid data struct")
+	}
+
+	// if this is not a control-plane node, this phase should not be executed
+	if !data.IsControlPlaneNode() {
```

Comment:
The preflight phase might also contain other checks. One notable check, that's missing and is applicable to both worker and control plane nodes is the root check.
Hence, we should probably not make this as central as it is and we should probably make the comment above state that.

Reply:
done. If this is a control-plane node, the pull the basic images. And run other check on both worker and control plane nodes.